Docs/fix lb multicomponent (#4)
Clemson-MSE/lammps: doc/src/fix_lb_multicomponent.rst

.. index:: fix lb/multicomponent

fix lb/multicomponent command
==============================

Syntax
""""""

.. parsed-literal::

* ID, group-ID are documented in :doc:`fix <fix>` command
* lb/multicomponent = style name of this fix command
* nevery = update the lattice-Boltzmann fluid every this many timesteps (should normally be 1)
* viscosity = the fluid viscosity (units of mass/(time\*length))

> * density = the initial fluid density

Reviewer: ```suggestion
* density = the fluid density
```

> * D3Q19 = velocity lattice model (required)

Reviewer: ```suggestion
* D3Q19 = the only velocity vector set used in this fix (mandatory to mention)
```

* dx = keyword for lattice spacing
* dx_value = the lattice-Boltzmann grid spacing
* zero or more keyword/value pairs may be appended
* keyword = *tau_r* or *tau_p* or *tau_s* or *kappa1* or *kappa2* or *kappa3* or *alpha* or *gamma_p* or *gamma_s* or *C1* or *C2* or *C3* or *dumpxdmf* or *seed* or *init*

  .. parsed-literal::

>        *tau_r* value = relaxation time for momentum transport (related to viscosity)

Reviewer: ```suggestion
       *tau_r* value = relaxation time constant for the population distribution *f* - related to viscosity
```

>        *tau_p* value = relaxation time for order parameter φ transport

Reviewer: ```suggestion
       *tau_p* value = relaxation time constant for the population distribution *g* - related to the mobility coefficient of the order parameter $\phi$
```

>        *tau_s* value = relaxation time for order parameter ψ transport

Reviewer: ```suggestion
       *tau_s* value = relaxation time constant for the population distribution *k* - related to the mobility coefficient of the order parameter $\psi$
```

>        *kappa1* value = surface tension parameter between components C1 and C2

Reviewer: ```suggestion
       *kappa1* value = energy gradient parameter for fluid 1 - used in surface tension calculation between fluid 1 and other fluids
```

>        *kappa2* value = surface tension parameter between components C1 and C3

Reviewer: ```suggestion
       *kappa2* value = energy gradient parameter for fluid 2 - used in surface tension calculation between fluid 2 and other fluids
```

>        *kappa3* value = surface tension parameter for component C3

Reviewer: ```suggestion
       *kappa3* value = energy gradient parameter for fluid 3 - used in surface tension calculation between fluid 3 and other fluids
```

>        *alpha* value = gradient energy coefficient

Reviewer: ```suggestion
       *alpha* value = parameter related to the interface width measurement
```

>        *gamma_p* value = mobility coefficient for order parameter φ

Reviewer: ```suggestion
       *gamma_p* value = mobility coefficient for order parameter $\phi$
```

>        *gamma_s* value = mobility coefficient for order parameter ψ

Reviewer: ```suggestion
       *gamma_s* value = mobility coefficient for order parameter $\psi$
```

>        *C1* value = bulk concentration of component 1

Reviewer: ```suggestion
       *C1* value = bulk concentration of fluid component 1
```

>        *C2* value = bulk concentration of component 2

Reviewer: ```suggestion
       *C2* value = bulk concentration of fluid component 2
```

>        *C3* value = bulk concentration of component 3 (note: C1 + C2 + C3 = 1)

Reviewer: ```suggestion
       *C3* value = bulk concentration of fluid component 3 (note: C1 + C2 + C3 = 1)
```

>        *dumpxdmf* values = N filename timeI

Reviewer: ```suggestion
       *dumpxdmf* values = N filename
```

           N = output fluid fields every N timesteps
           filename = base name for output files (.xdmf and .raw extensions added automatically)

>            timeI = 1 (use simulation time to index xdmf file), 0 (use output frame number)

Reviewer: ```suggestion
```

>        *seed* value = random number generator seed (positive integer)

Reviewer: ```suggestion
       *seed* value = random number generator seed (positive integer) - useful to initialize certain models like ternary mixture, ternary film, and ternary droplet 
```

       *init* values = initialization_method [method_parameters]
           initialization_method = *mixture* or *droplet* or *liquid_lens* or *double_emulsion* or *film* or *mixed_droplet*

>                *mixture* = homogeneous mixture with random fluctuations

Reviewer: ```suggestion
               *mixture* = well-mixed fluid mixture
```

>                *droplet* radius = binary droplet (C1 and C2) of given radius in C3 solvent

Reviewer: ```suggestion
               *droplet* radius = droplet in a solvent model - droplet of fluid component 1 of given radius surrounded by fluid component 2 - droplet center is located at the center of the box
```

               *liquid_lens* radius = lens of C3 with given radius between C1 and C2 layers
               *double_emulsion* radius = Janus droplet (C1 and C2 hemispheres) of given radius in C3 solvent
               *film* thickness C1_film C2_film = ternary film with specified thickness and composition
               *mixed_droplet* radius C1_drop C2_drop = ternary droplet with specified radius and composition in bulk

Examples
""""""""

.. code-block:: LAMMPS

   fix 1 all lb/multicomponent 1 0.166667 1.0 D3Q19 dx 1.0 init mixture
   fix 2 fluid lb/multicomponent 1 0.166667 1.0 D3Q19 dx 1.0 init droplet 10.0
   fix 3 all lb/multicomponent 1 0.166667 1.0 D3Q19 dx 1.0 C1 0.333333 C2 0.333333 C3 0.333334 kappa1 0.01 kappa2 0.02 kappa3 0.05 init mixture dumpxdmf 1000 output 0

Description
"""""""""""

.. versionadded:: 2024

> The **fix lb/multicomponent** command implements a ternary lattice Boltzmann model (LBM) for simulating multicomponent fluid mixtures in three dimensions. This fix is an extension of the single-component :doc:`fix lb/fluid <fix_lb_fluid>`, designed to capture more complex interfacial and phase behaviors such as droplets, emulsions, and films involving three interacting fluid components.

Reviewer: ```suggestion
The **fix lb/multicomponent** command implements a ternary lattice Boltzmann model (LBM) for simulating three-dimensional multicomponent fluid systems. This fix is an extension of the single-component :doc:`fix lb/fluid <fix_lb_fluid>`, designed to capture complex interfacial and phase behaviors of three immiscible fluid components for various fluid models like ternary mixture, droplets, and films.
```

> The algorithm evolves three coupled distribution functions over a D3Q19 lattice and models thermodynamic interactions via chemical potentials and free-energy gradients. The model is suitable for studying phase separation, interface dynamics, droplet stability, and multiphase coexistence.

Reviewer: ```suggestion
The algorithm evolves three distribution functions over a D3Q19 lattice and models thermodynamic interactions using the Landau-Ginzburg free-energy. This model is suitable for studying multi-component fluid interactions in the aspect of phase separation and interface dynamics.
```

The implementation follows the work described in Arumugam Kumar et al., "Implementation of a Ternary Lattice Boltzmann Model in LAMMPS," Comput. Phys. Commun. 294, 108898 (2024).

----------

**Model Overview**

> The ternary fluid composition is specified by three concentration fields C₁, C₂, and C₃, which are related to conserved order parameters:

Reviewer: ```suggestion
The ternary fluid composition is specified by three concentration fields $C_1$, $C_2$, and $C_3$, which are related to conserved order parameters:
```

.. math::

   \rho &= C_1 + C_2 + C_3 \\
   \phi &= C_1 - C_2 \\
   \psi &= C_3

> where ρ is the total density (approximately constant for incompressible flow), and φ and ψ are order parameters that distinguish the three components.

Reviewer: ```suggestion
where $\rho$ is the total density (approximately constant for incompressible flow), $\phi$ and \psi$ are order parameters that distinguish the three components.
```

The time evolution is described by three sets of distribution functions (f, g, h) that evolve according to the Bhatnagar-Gross-Krook (BGK) lattice Boltzmann equation:

.. math::

   ```

>    \rho &= C_1 + C_2 + C_3 = 1 \\

Reviewer: change to the LBM time evolution equations for f, g, h and remove rho, phi, psi

   \phi &= C_1 - C_2 \\
   \psi &= C_3
   ```

> with :math:`C_1, C_2, C_3` denoting component concentrations.

Reviewer: ```suggestion
with $\tau_\rho, $\tau_\phi$, and $\tau_psi$ denoting the time constants for individual LBM equations, and has relations to macroscopic properties of viscosity and mobility coefficients.
```

> with :math:`C_1, C_2, C_3` denoting component concentrations.

Reviewer: Include equations talking about time constants and how they are related to fluid properties like viscosity and mobility coefficients.

## Keywords

----------

**Initialization Methods**

The *init* keyword specifies how the ternary fluid is initialized:

> * **mixture**: Creates a homogeneous mixture with bulk concentrations C1, C2, C3 plus small random fluctuations (±0.01) to seed phase separation. This is the default initialization if no *init* keyword is specified.

Reviewer: ```suggestion
* **mixture**: Creates a fluid mixture with bulk concentrations C1, C2, C3. This is the default initialization if no *init* keyword is specified.
```

> * **droplet** *radius*: Creates a binary droplet composed of equal parts C1 and C2 (C3 = 0) with specified radius (in lattice units) centered in the simulation box, surrounded by pure C3.

Reviewer: ```suggestion
* **droplet** *radius*: Creates a binary droplet composed of components 1 and 2 with concentration C1 and C2 (C1+C2=1) with specified radius (in lattice units) centered in the simulation box, surrounded by component 3 in bulk.
```

> * **liquid_lens** *radius*: Creates a lens-shaped droplet of pure C3 with specified radius positioned at the interface between horizontal layers of pure C1 (top half) and pure C2 (bottom half).

Reviewer: ```suggestion
* **liquid_lens** *radius*: Creates a lens-shaped droplet of component 3 with equal upper and lower radius of curvature, positioned at the interface between horizontal layers of component 1 (top half) and component 2 (bottom half).
```

> * **double_emulsion** *radius*: Creates a Janus-type double emulsion with a droplet of specified radius split into two hemispheres: left hemisphere of pure C1 and right hemisphere of pure C2, surrounded by pure C3 solvent.

Reviewer: ```suggestion
* **double_emulsion** *radius*: Creates a Janus-type double emulsion with a droplet of specified radius split into two hemispheres: left hemisphere of component 1 and right hemisphere of component 2, centered in the simulation box and surrounded by component 3 in bulk.
```

> * **film** *thickness* *C1_film* *C2_film*: Creates a horizontal film of specified thickness (as a fraction of the box height) centered in the simulation box with composition C1_film, C2_film, C3_film = 1 - C1_film - C2_film, surrounded by bulk mixture. Small random fluctuations (±0.01) are added to seed phase separation.

Reviewer: ```suggestion
* **film** *thickness* *C1_film* *C2_film*: Creates a horizontal film of specified thickness (as a fraction of the box height) centered in the simulation box with composition C1_film, C2_film, C3_film (C1_film + C2_film + C3_film = 1), surrounded by a bulk mixture of C1, C2, C3 (C1+C2+C3 = 1).
```

Reviewer: We can phrase "Small random fluctuations (±0.01) are added to seed phase separation" a bit differently, without being technically incorrect. Need to discuss.

> * **mixed_droplet** *radius* *C1_drop* *C2_drop*: Creates a ternary droplet with specified radius and composition (C1_drop, C2_drop, C3_drop = 1 - C1_drop - C2_drop) surrounded by pure C3 solvent. Small random fluctuations (±0.01) are added within the droplet.

Reviewer: ```suggestion
* **mixed_droplet** *radius* *C1_drop* *C2_drop*: Creates a ternary droplet with specified radius and composition (C1_drop, C2_drop, C3_drop = 1 - C1_drop - C2_drop) surrounded by component 3 in bulk.
```

> .. math::

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

>    \text{subNbx} = \frac{N_x}{P_x} + 2h_x \\

Reviewer: ```suggestion
```

>    \text{subNby} = \frac{N_y}{P_y} + 2h_y \\

Reviewer: ```suggestion
```

>    \text{subNbz} = \frac{N_z}{P_z} + 2h_z

Reviewer: ```suggestion
```

> where hₓ = hᵧ = hᵤ = 2 is the halo extent required for gradient and Laplacian calculations.

Reviewer: ```suggestion
```

**Parameters**

This fix accepts several internal parameters that control the thermodynamic and transport properties:

> * **tau_r**, **tau_p**, **tau_s**: Relaxation times for momentum transport and order parameter transport. These control the viscosity and diffusivity of the fluid components.

Reviewer: ```suggestion
* **tau_r**, **tau_p**, **tau_s**: Relaxation times for momentum transport and order parameter transport.
```

> * **gamma_p**, **gamma_s**: Mobility coefficients controlling the rate of change of the phase variables φ and ψ.

Reviewer: ```suggestion
* **gamma_p**, **gamma_s**: Parameters controlling the mobility of the phases $\phi$ and $\psi$.
```

* **kappa1**, **kappa2**, **kappa3**: Surface tension parameters that determine the interfacial tensions between the three components. The surface tension between components m and n is given by:

  .. math::

>      \gamma_{mn} = \frac{\sqrt{(\kappa_m + \kappa_n)(\lambda_m + \lambda_n)}}{6}

Reviewer: ```suggestion
     \gamma_{mn} = \frac{\alpha}{6}\left(\kappa_m + \kappa_n\right)
```

> * **alpha**: Gradient energy coefficient.

Reviewer: ```suggestion
* **alpha**: Parameter correlated with the interface width in LB units. The interface width is given by 4*alpha.
```

> * **C1**, **C2**, **C3**: Bulk concentrations of the three components (must sum to 1.0).

Reviewer: ```suggestion
* **C1**, **C2**, **C3**: Bulk concentrations of the three components.
```

Future versions may allow passing these parameters directly via the input script.

----------

**Output**

> The *dumpxdmf* keyword enables output of the fluid fields to files that can be visualized using Paraview or other XDMF-compatible visualization software. The output includes:

Reviewer: ```suggestion
The *dumpxdmf* keyword enables output of the fluid fields to files that can be visualized using Paraview or other XDMF-compatible visualization software. The output (in LB units) includes:
```

> * Density field ρ (converted to physical units)

Reviewer: ```suggestion
* Density field ρ
```

> * Order parameters φ and ψ (converted to physical units)

Reviewer: ```suggestion
* Order parameters \phi and \psi
```

> * Pressure field p (converted to physical units)

Reviewer: ```suggestion
* Pressure field p
```

> * Velocity field **u** (converted to physical units)

Reviewer: ```suggestion
* Velocity vector field **u**
```

Two files are created:

* A text .xdmf file containing metadata and grid structure
* A binary .raw file containing the actual field data

> The concentration fields can be recovered from the output using:

Reviewer: ```suggestion
```

> .. math::

Reviewer: ```suggestion
```

>    C_1 = \frac{\rho + \phi}{2}, \quad C_2 = \frac{\rho - \phi}{2}, \quad C_3 = \psi

Reviewer: ```suggestion
```

----------

**Boundary Conditions**

> The ternary lattice-Boltzmann implementation currently supports periodic boundary conditions in all three directions. The boundary conditions are specified through the main LAMMPS :doc:`boundary <boundary>` command and must be set to periodic (p p p) for proper operation.

Reviewer: ```suggestion
The ternary lattice-Boltzmann implementation currently only supports periodic boundary conditions in all three directions. The boundary conditions are specified through the main LAMMPS :doc:`boundary <boundary>` command and must be set to periodic (p p p) for proper operation.
```

> *Periodic Boundaries*

Reviewer: ```suggestion
```

> With periodic boundary conditions, the fluid domain wraps around in all directions. The distribution functions that stream out of one side of the simulation box re-enter from the opposite side. This is handled automatically through the MPI communication scheme, which includes a halo region of 2 lattice layers on each side of the local processor subdomain.

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> The halo communication ensures that:

Reviewer: ```suggestion
```

> * Gradient calculations (∇ρ, ∇φ, ∇ψ) required for the chemical potentials can be performed correctly near subdomain boundaries

Reviewer: ```suggestion
```

> * Laplacian calculations (∇²ρ, ∇²φ, ∇²ψ) have access to necessary neighbor information

Reviewer: ```suggestion
```

> * Streaming of distribution functions across processor boundaries is handled correctly

Reviewer: ```suggestion
```
